pep-9999.rst

```diff
@@ -0,0 +1,129 @@
+PEP: 9999
+Title: Platform Tag for Linux Distributions Using Musl
+Author: Tzu-ping Chung <uranusjr@gmail.com>
+Sponsor: Brett Cannon <brett@python.org>
+PEP-Delegate: TBD
+Discussions-To: https://discuss.python.org/t/7165
+Status: Draft
+Type: Informational
+Content-Type: text/x-rst
+Created: 2021-03-17
+
+
+Abstract
+========
+
+This PEP proposes a new platfrom tag series ``musllinux`` for
+binary Python package distributions for a Python installation linked
+against musl on a Linux distribution. The tag works similarly to the
+"perennial manylinux" platform tags specified in :pep:`600`, but
+targeting platforms based on musl instead.
+
+
+Motivation
+==========
+
+With the wide use of containers, distributions such as Alpine Linux,
+[alpine]_ have been gaining more popularity than ever. Many of them
+based on musl, [musl]_ a different libc implementation from glibc, and
+therefore cannot use the existing ``manylinux`` platform tags. This
+means that Python package projects cannot deploy binary distributions
+on PyPI for them. Users of such projects demand build constraints from
+those projects, putting unnecessary burden on project maintainers.
+
+
+Rationale
+=========
+
+According to the documentation, musl has a stable ABI, and maintains
+backwards compatibility, [musl-compatibility]_ [compare-libcs]_ so a
+binary compiled against an earlier version of musl is guaranteed to
+run against a newer musl runtime. [musl-compat-ml]_ Therefore, we use
+a scheme similar to the glibc-version-based manylinux tags, but
+against musl versions instead of glibc.
+
+Logic behind the new platform tag largely follows :pep:`600`, and
+require wheels using this tag make similar promises. Please refer to
+the PEP for more details on rationale and reasoning behind the design.
+
+
+Specification
+=============
+
+Tags using the new scheme will take the form::
+
+    musllinux_${MUSLMAJOR}_${MUSLMINOR}_${ARCH}
+
+Distributions using the tag make similar promises to those discribed
+in :pep:`600`, including:
+
+1. The distribution works on any mainstream Linux distributions with
+   musl version ``${MUSLMAJOR}.${MUSLMINOR}`` or later.
+2. The distribution's ``${ARCH}`` matches the return value of
+   ``sysconfig.get_platform()`` on the host system.
+
```

Comment:
How about some examples?

Alpine 3.13 contains with ``musl-1.2.2-r0``. A tag for an AMD64 build be ``musllinux_1_2_x86_64``. A tag for Alpine 3.12 on ARM64 would be ``musllinux_1_1_aarch64`` (unless I got the platform tag wrong).

Reply:
PS: Alpine 3.1 to 3.12 seems to be musl-1.1.x.

Reply:
The format I’m having in mind is e.g.

```
musllinux_1_2_x86_64
musllinux_1_2_aarch64  # Probably, I don't have an ARM machine at hand to check.
```

I’ll add some examples and maybe example code to make it clearer what’s expected. The logic should be

```python
import sysconfig

def format_musllinux(musl_version: tuple[int, int]) -> str:
    plat = sysconfig.get_platform()
    assert plat[:6] == "linux-", "Not a Linux"
    host = plat.split("-", 1)[-1].replace("-", "_")
    return f"musllinux_{musl_version[0]}_{musl_version[1]}_{host}"
```